In [1]:
import os
import sys
ROOT = os.path.abspath(os.path.join(os.getcwd(), os.pardir))
SRC = os.path.abspath(os.path.join(ROOT, "src"))
sys.path.append(ROOT)
sys.path.append(SRC)

In [2]:
import dvc.api
import pandas as pd
from src.etl.transform import change_columns_types
from src.etl.extract import extract_data
params = dvc.api.params_show()
types_columns: list[str] = params['preparation']['types']['columns']


file_prepared_data =os.path.join(ROOT, params['paths']['data'])
df = extract_data(file_prepared_data)
# df = change_columns_types(df, types_columns)

### Description of variables

In [3]:
typed_columns = params['preparation']['types']['columns']

typed_columns

{'InvoiceNo': 'str',
 'StockCode': 'str',
 'Description': 'str',
 'Quantity': 'int',
 'InvoiceDate': 'datetime',
 'UnitPrice': 'float',
 'CustomerID': 'str',
 'Country': 'str'}

In [4]:
df.shape

(541909, 8)

### Missing values

In [5]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [6]:
for col, typ in typed_columns.items():
    print("======")
    print(col, typ)
    print(df[col].describe())
    print("******")

InvoiceNo str
count     541909
unique     25900
top       573585
freq        1114
Name: InvoiceNo, dtype: object
******
StockCode str
count     541909
unique      4070
top       85123A
freq        2313
Name: StockCode, dtype: object
******
Description str
count                                 540455
unique                                  4223
top       WHITE HANGING HEART T-LIGHT HOLDER
freq                                    2369
Name: Description, dtype: object
******
Quantity int
count    541909.000000
mean          9.552250
std         218.081158
min      -80995.000000
25%           1.000000
50%           3.000000
75%          10.000000
max       80995.000000
Name: Quantity, dtype: float64
******
InvoiceDate datetime
count               541909
unique               23260
top       10/31/2011 14:41
freq                  1114
Name: InvoiceDate, dtype: object
******
UnitPrice float
count    541909.000000
mean          4.611114
std          96.759853
min      -11062.060000
25%         

In [66]:
negative_quantity = df[df['Quantity'] < 0 ]
negative_quantity['Quantity'].describe()

count    10624.000000
mean       -45.607210
std       1092.214216
min     -80995.000000
25%        -10.000000
50%         -2.000000
75%         -1.000000
max         -1.000000
Name: Quantity, dtype: float64

In [85]:
negative_quantity = df[df['Quantity'] > 0 ]
negative_quantity['Quantity'].describe()

count    531285.000000
mean         10.655262
std         156.830323
min           1.000000
25%           1.000000
50%           3.000000
75%          10.000000
max       80995.000000
Name: Quantity, dtype: float64

In [84]:
negative_quantity[negative_quantity['Quantity'] < -10 ].sort_values(by='Quantity')

InvoiceNo StockCode                          Description  Quantity  \
540422   C581484     23843          PAPER CRAFT , LITTLE BIRDIE    -80995   
61624    C541433     23166       MEDIUM CERAMIC TOP STORAGE JAR    -74215   
225530    556691     23005         printing smudges/thrown away     -9600   
225529    556690     23005         printing smudges/thrown away     -9600   
4287     C536757     84347  ROTATING SILVER ANGELS T-LIGHT HLDR     -9360   
...          ...       ...                                  ...       ...   
37879    C539486     22603        CHRISTMAS RETROSPOT TREE WOOD       -11   
150605    549466     20816                                  NaN       -11   
198246    554024     20998                                  NaN       -11   
279256    561254  DCGS0067                                 ebay       -11   
150151    549340     20774                                  NaN       -11   

            InvoiceDate  UnitPrice  CustomerID         Country  
540422   12/9/2011 9:27       2.08     16446.0  United Kingdom  
61624   1/18/2011 10:17       1.04     12346.0  United Kingdom  
225530  6/14/2011 10:37       0.00         NaN  United Kingdom  
225529  6/14/2011 10:37       0.00         NaN  United Kingdom  
4287    12/2/2010 14:23       0.03     15838.0  United Kingdom  
...                 ...        ...         ...             ...  
37879   12/20/2010 8:27       0.85     18256.0  United Kingdom  
150605   4/8/2011 14:22       0.00         NaN  United Kingdom  
198246  5/20/2011 13:21       0.00         NaN  United Kingdom  
279256  7/26/2011 11:52       0.00         NaN  United Kingdom  
150151   4/8/2011 11:02       0.00         NaN  United Kingdom  

[2610 rows x 8 columns]

In [7]:
df[df['CustomerID'] == 13136.0]

# df[(df['StockCode'] == '22633') && (df['CustomerID'] == 13136.0])]

# df[df['InvoiceNo'] == '571224']

InvoiceNo StockCode                         Description  Quantity  \
21748    C538119     22633              HAND WARMER UNION JACK        -1   
21749    C538119     22865              HAND WARMER OWL DESIGN        -1   
21750    C538119     21777         RECIPE BOX WITH METAL HEART        -2   
185445    552817     23298                      SPOTTY BUNTING         6   
185446    552817     47566                       PARTY BUNTING         8   
...          ...       ...                                 ...       ...   
398175    571224     23389              SPACEBOY MINI BACKPACK         4   
410567   C572111     23437       50'S CHRISTMAS GIFT BAG LARGE        -2   
410568   C572111     82483  WOOD 2 DRAWER CABINET WHITE FINISH        -1   
410569   C572111     22186                RED STAR CARD HOLDER        -3   
410570   C572111     22138       BAKING SET 9 PIECE RETROSPOT         -1   

             InvoiceDate  UnitPrice  CustomerID         Country  
21748    12/9/2010 15:34       2.10     13136.0  United Kingdom  
21749    12/9/2010 15:34       2.10     13136.0  United Kingdom  
21750    12/9/2010 15:34       7.95     13136.0  United Kingdom  
185445   5/11/2011 12:47       4.95     13136.0  United Kingdom  
185446   5/11/2011 12:47       4.95     13136.0  United Kingdom  
...                  ...        ...         ...             ...  
398175  10/14/2011 13:41       4.15     13136.0  United Kingdom  
410567  10/20/2011 19:13       1.25     13136.0  United Kingdom  
410568  10/20/2011 19:13       6.95     13136.0  United Kingdom  
410569  10/20/2011 19:13       1.95     13136.0  United Kingdom  
410570  10/20/2011 19:13       4.95     13136.0  United Kingdom  

[139 rows x 8 columns]

In [8]:
# Checking when Quantity is negative

df[df['Quantity'] == -2 ].sort_values(by='Description')

# Some represents 
# - Discounts 

InvoiceNo StockCode                     Description  Quantity  \
410567   C572111     23437   50'S CHRISTMAS GIFT BAG LARGE        -2   
240319   C558097     22167      OVAL WALL MIRROR DIAMANTE         -2   
431997   C573779     23007          SPACEBOY BABY GIFT SET        -2   
466260   C576324     20973   12 PENCIL SMALL TUBE WOODLAND        -2   
20527    C538062     20973   12 PENCIL SMALL TUBE WOODLAND        -2   
...          ...       ...                             ...       ...   
453963    575503     35933                             NaN        -2   
468300    576415    17012B                             NaN        -2   
473440    576835     21618                             NaN        -2   
517719    580087    84341B                             NaN        -2   
535322    581199     84581                             NaN        -2   

             InvoiceDate  UnitPrice  CustomerID         Country  
410567  10/20/2011 19:13       1.25     13136.0  United Kingdom  
240319   6/26/2011 13:29       9.95     15532.0  United Kingdom  
431997   11/1/2011 10:57      16.95     16360.0  United Kingdom  
466260  11/14/2011 15:10       0.65     16161.0  United Kingdom  
20527    12/9/2010 13:40       0.65     17702.0  United Kingdom  
...                  ...        ...         ...             ...  
453963  11/10/2011 10:19       0.00         NaN  United Kingdom  
468300  11/15/2011 11:22       0.00         NaN  United Kingdom  
473440  11/16/2011 15:10       0.00         NaN  United Kingdom  
517719   12/1/2011 14:27       0.00         NaN  United Kingdom  
535322   12/7/2011 18:26       0.00         NaN  United Kingdom  

[1395 rows x 8 columns]

In [9]:
# Group by InvoiceNo.
invoice_ids_with_only_negative_quantity = df.groupby('InvoiceNo').filter(lambda g: (g['Quantity'] < 0).all())['InvoiceNo'].unique()
df_with_valid_invoices = df[~df['InvoiceNo'].isin(invoice_ids_with_only_negative_quantity)]

In [11]:
# Checking description with nan
df_with_valid_invoices['Quantity'].describe()

count    531285.000000
mean         10.655262
std         156.830323
min           1.000000
25%           1.000000
50%           3.000000
75%          10.000000
max       80995.000000
Name: Quantity, dtype: float64

In [12]:
df_with_valid_invoices[df_with_valid_invoices['Description'].isna()]['UnitPrice'].describe()

count    592.0
mean       0.0
std        0.0
min        0.0
25%        0.0
50%        0.0
75%        0.0
max        0.0
Name: UnitPrice, dtype: float64

In [41]:
df_semi_cleaned = df_with_valid_invoices[~(df_with_valid_invoices['Description'].isna() & (df_with_valid_invoices['UnitPrice'] == 0))]
df_semi_cleaned[df_semi_cleaned['UnitPrice'] == 0]

print(df_semi_cleaned.shape)
print(df_semi_cleaned.columns)


invoices = df_semi_cleaned['InvoiceNo'].unique()

# b = df_semi_cleaned.groupby('InvoiceNo').apply(lambda x: x)
# print(b.shape)
# print(b.columns)

(530693, 8)
Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')


In [4]:
import src.utils

from etl.extract import check_existing_columns, drop_empty_invoice_id, drop_unit_price_less_than_zero, extract_data, filter_invoices_with_non_negative_quantity, filter_shop_tiems_with_empty_description_and_zero_unit_price
from etl.transform import change_columns_types, sample_dataframe

from preparation import process

process_df = process(df)

In [7]:
process_df.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

Summary:
- The database is a time serie about shopping stream.
- The first record is on December 2010. The duration of the record is one year.

Assumptions:
- There is a relation between CustomerId and InvoiceId. One custuomber may have more of one Invoice.
- A Invoice can contain more of one record (each entry in the db is a shop record)
- Quantity and UnitPrice are columns with positive values.
- InvoiceId is a primary key from antother table. 

Problems:
1) There are some items with empty CustomerId, but InvoiceID is well formed. This cases represents around 20 %
2) There are cases with Description is empty. This case is around the 1454 / 541909 * 100 ≃ 0.002 percent of the data.
3) There are ~10 K cases when Quantify is negative. Some cases represents discounts.
4) There are cases when UnitPice is lower to zero
   

Fixes:
1) Fill empty CustomerId checking the relates CustomerId from InvoiceId. This process is like a group by Invoice Id. Apply the following rules:
   - Delete rows when CustomerId is empty or nan.
   - Groupded by InvoiceId.
       - If exists more than one CustomerID, then ignored
       - In another case, fill te CustomerId with that Customer
2) Delete cases when Description is Empty
3) Some cases represents discounts. Follow the next rules:
   - If a invoice only contains shops items with negative unit prices, then is a refund case. In this cases, drop that invoice.
   - Then, a valid Invoice for this model, only ocurrs when a Invoice contains neative or positive values. 